From e871cc9c33303e212b2ef1ab5e49b31f37c593eb Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Thu, 20 Apr 2023 18:11:04 +0800
Subject: [PATCH 0106/1646] [Huawei] mm/dynamic_hugetlb: fix type error of pfn
 in __hpool_split_gigantic_page()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 46904

--------------------------------

The type of pfn is int, which can result in truncation.
Change its type to unsigned long to fix the problem.

Fixes: eef7b4fd04a0 ("mm/dynamic_hugetlb: use pfn to traverse subpages")
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 mm/dynamic_hugetlb.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 526d84870a60..8199ef893f4a 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -55,7 +55,8 @@ static void __hpool_split_gigantic_page(struct dhugetlb_pool *hpool, struct page
 {
 	int nr_pages = 1 << (PUD_SHIFT - PAGE_SHIFT);
 	int nr_blocks = 1 << (PMD_SHIFT - PAGE_SHIFT);
-	int i, pfn = page_to_pfn(page);
+	unsigned long pfn = page_to_pfn(page);
+	int i;
 
 	lockdep_assert_held(&hpool->lock);
 	atomic_set(compound_mapcount_ptr(page), 0);
-- 
2.17.1

